src/deque.rs

```diff
@@ -0,0 +1,775 @@
+                self.done = true;
+            }
+            Some(unsafe { &*(self.buffer.add(self.back) as *const T) })
+        }
+    }
+}
+
+impl<'a, T, const N: usize> ExactSizeIterator for Iter<'a, T, N> {}
+impl<'a, T, const N: usize> FusedIterator for Iter<'a, T, N> {}
+
+/// An iterator over the elements of a [`Deque`].
+///
+/// This struct is created by calling the `iter` method.
+pub struct IterMut<'a, T, const N: usize> {
+    buffer: *mut MaybeUninit<T>,
+    _phantom: PhantomData<&'a mut T>,
+    front: usize,
+    back: usize,
+    done: bool,
+}
+
+impl<'a, T, const N: usize> Iterator for IterMut<'a, T, N> {
+    type Item = &'a mut T;
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.done {
+            None
+        } else {
+            let index = self.front;
+            self.front = Deque::<T, N>::increment(self.front);
+            if self.front == self.back {
+                self.done = true;
+            }
+            Some(unsafe { &mut *(self.buffer.add(index) as *mut T) })
+        }
+    }
+
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        let len = if self.done {
+            0
+        } else if self.back <= self.front {
+            self.back + N - self.front
+        } else {
+            self.back - self.front
+        };
+
+        (len, Some(len))
+    }
+}
+
+impl<'a, T, const N: usize> DoubleEndedIterator for IterMut<'a, T, N> {
+    fn next_back(&mut self) -> Option<Self::Item> {
+        if self.done {
+            None
+        } else {
+            self.back = Deque::<T, N>::decrement(self.back);
+            if self.front == self.back {
+                self.done = true;
+            }
+            Some(unsafe { &mut *(self.buffer.add(self.back) as *mut T) })
+        }
+    }
+}
+
+impl<'a, T, const N: usize> ExactSizeIterator for IterMut<'a, T, N> {}
+impl<'a, T, const N: usize> FusedIterator for IterMut<'a, T, N> {}
+
+impl<'a, T, const N: usize> IntoIterator for &'a Deque<T, N> {
+    type Item = &'a T;
+    type IntoIter = Iter<'a, T, N>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        self.iter()
+    }
+}
+
+impl<'a, T, const N: usize> IntoIterator for &'a mut Deque<T, N> {
+    type Item = &'a mut T;
+    type IntoIter = IterMut<'a, T, N>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        self.iter_mut()
+    }
+}
+
+impl<T, const N: usize> Clone for Deque<T, N>
+where
+    T: Clone,
+{
+    fn clone(&self) -> Self {
+        let mut res = Deque::new();
+        for i in self {
+            // safety: the original and new deques have the same capacity, so it can
+            // not become full.
+            unsafe { res.push_back_unchecked(i.clone()) }
+        }
+        res
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::Deque;
+
+    #[test]
+    fn static_new() {
+        static mut _V: Deque<i32, 4> = Deque::new();
+    }
+
+    #[test]
+    fn stack_new() {
+        static mut _V: Deque<i32, 4> = Deque::new();
+    }
```

Comment:
copy-paste error ?

Reply:
It's also wrong in `Vec` :rofl:. Fixed in both.